Cleanup use of pthreads API

peter-b opened the pull request:
The `pthread_t` type is opaque, and its structure is
implementation-defined.  The POSIX thread specification doesn't
provide any value for a `pthread_t` that is guaranteed to be an
invalid thread identifier, so it can't be validly initialised to
`NULL` or compared with `NULL`.

Similarly, `pthread_t` values can only be safely compared with
`pthread_equal()`, not with `==`.

The new `MCManagedPThread` class provides an implicit `bool` cast
operator so that it can be used directly in `if` statements, provides
`==` operators for other `MCManagedPThread` values and for `pthread_t`
values for comparison purposes, and provides `Join()` and `Create()`
methods that wrap the corresponding pthread API procedures.

Most uses of `pthread_t` and related procedure calls can be replaced with `MCManagedPThread` instances.

peter-b committed:
Use a MCManagedPThread for the socket polling thread.

peter-b committed:
Use MCManagedPThread in MCFiber.

peter-b committed:
Compare pthread_t values using pthread_equal().

livecodefraser commented:
@livecode-vulcan review okay dbfd95d

livecode-vulcan commented:
💙 review by @livecodefraser ok dbfd95d6e2483bbe12ca67d4ef468824b77516eb

livecode-vulcan commented:
😎 test success dbfd95d6e2483bbe12ca67d4ef468824b77516eb
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2740)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2750)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2735)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2751)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2757)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2763)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2775)

livecodefraser commented:
@livecode-vulcan review okay 64f3483

livecode-vulcan commented:
💙 review by @livecodefraser ok 64f3483e46bd6b29159981bbdb654658e2a58ea7

livecodefraser commented:
@livecode-vulcan review okay c810fe1

livecode-vulcan commented:
💙 review by @livecodefraser ok c810fe195eaa95863e5e29c4b3ca90409565f6c8

livecode-vulcan commented:
😞 test failure c810fe195eaa95863e5e29c4b3ca90409565f6c8
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2746)
- try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/2755)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2741)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2757)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/2763)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2767)
- try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2784)

livecodefraser commented:
@livecode-vulcan review okay a3e9ff9

livecode-vulcan commented:
💙 review by @livecodefraser ok a3e9ff97567ca63e07ba050629a95cd129f8ebbb

livecode-vulcan commented:
😎 test success a3e9ff97567ca63e07ba050629a95cd129f8ebbb
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2747)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2756)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2742)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2758)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2764)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2768)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2786)

peter-b committed:
Add MCManagedPThread class.

Add a helper class for managing a `pthread_t` value and associated
"liveness" state.

The `pthread_t` type is opaque, and its structure is
implementation-defined.  The POSIX thread specification doesn't
provide any value for a `pthread_t` that is guaranteed to be an
invalid thread identifier, so it can't be validly initialised to
`NULL` or compared with `NULL`.

Similarly, `pthread_t` values can only be safely compared with
`pthread_equal()`, not with `==`.

The new `MCManagedPThread` class provides an implicit `bool` cast
operator so that it can be used directly in `if` statements, provides
`==` operators for other `MCManagedPThread` values and for `pthread_t`
values for comparison purposes, and provides `Join()` and `Create()`
methods that wrap the corresponding pthread API procedures.

peter-b committed:
Use MCManagedPThreads for the Android UI/engine threads.

peter-b commented:
@livecodefraser I've tracked down the problem.  I mistakenly thought that the `s_socket_poll_thread_run` variable was to indicate whether the socket poll thread was running; it was actually being used to signal the thread to tell it to exit.  I've changed the name of the variable and made it explicitly a `sig_atomic_t` in order to make it clearer that it's being used for inter-thread communication.

I think it's okay for consumers of the `MCManagedPThread` to have to serialise access extrinsically, because there will likely be mutexes already being used for inter-thread coordination.

I've reviewed the changes very carefully again, and I don't think there should be any further issues with this patch.

livecodefraser commented:
@livecode-vulcan review okay c66680e

peter-b commented:
@livecode-vulcan review by @livecodefraser ok bbe2452

livecode-vulcan commented:
💙 review by @livecodefraser ok bbe2452721c4502647207dbbd08086bd480dcdde

livecode-vulcan commented:
😎 test success bbe2452721c4502647207dbbd08086bd480dcdde
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/128)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/128)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/128)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/128)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/128)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/128)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/128)
